satpy/readers/fci_l2_nc.py

```diff
@@ -291,24 +321,99 @@ def get_dataset(self, dataset_id, dataset_info):
+            self._area_def = self._construct_area_def(dataset_id)
 
-        float_variable = variable.where(variable != fill_value, mask_value).astype('float32', copy=False)
-        float_variable.attrs = variable.attrs
-        variable = float_variable
+            # coordinates are not relevant when returning data with an AreaDefinition
+            if 'coordinates' in dataset_info.keys():
+                del dataset_info['coordinates']
 
-        # Rename the dimensions as required by Satpy
-        variable = variable.rename({"number_of_FoR_rows": 'y', "number_of_FoR_cols": 'x'})
-#       # Manage the attributes of the dataset
-        variable.attrs.setdefault('units', None)
+        if 'fill_value' in dataset_info:
+            variable = self._mask_data(variable, dataset_info['fill_value'])
 
-        variable.attrs.update(dataset_info)
-        variable.attrs.update(self._get_global_attributes())
+        variable = self._set_attributes(variable, dataset_info, segmented=True)
+
+        return variable
+
+    def _construct_area_def(self, dataset_id):
+        """Construct the area definition.
+
+        Returns:
+            AreaDefinition: A pyresample AreaDefinition object containing the area definition.
+
+        """
+        res = dataset_id.resolution
+
+        area_naming_input_dict = {'platform_name': 'mtg',
+                                  'instrument_name': 'fci',
+                                  'resolution': res,
+                                  }
+
+        area_naming = get_geos_area_naming({**area_naming_input_dict,
+                                            **get_service_mode('fci', self.ssp_lon)})
+
+        # Construct area definition from standardized area definition.
+        stand_area_def = get_area_def(area_naming['area_id'])
+
+        if (stand_area_def.x_size != self.ncols) | (stand_area_def.y_size != self.nlines):
+            raise NotImplementedError('Unrecognised AreaDefinition.')
+
+        mod_area_extent = self._modify_area_extent(stand_area_def.area_extent)
+
+        area_def = geometry.AreaDefinition(
+            stand_area_def.area_id,
+            stand_area_def.description,
+            "",
+            stand_area_def.proj_dict,
+            stand_area_def.x_size,
+            stand_area_def.y_size,
+            mod_area_extent)
+
+        return area_def
+
+    @staticmethod
+    def _modify_area_extent(stand_area_extent):
+        """Modify area extent to macth satellite projection.
+
+        Area extent has to be modified since the L2 products are stored with the south-east
+        in the upper-right corner (as opposed to north-east in the standardized area definitions).
+        """
+        ll_x, ll_y, ur_x, ur_y = stand_area_extent
+        ll_y *= -1.
+        ur_y *= -1.
+        area_extent = tuple([ll_x, ll_y, ur_x, ur_y])
+
+        return area_extent
+
+    @staticmethod
+    def _slice_dataset(variable, dataset_info):
```

Comment:
these are four very similar pieces of code. Did you consider making that a function?

Reply:
Or maybe a loop?